File: miner/src/py/miner.py
import asyncio
import queue
import copy
from asyncio import StreamReader, StreamWriter
from threading import Thread, Lock, Event
from typing import Optional
from lib.src.types.py.crypto import PublicKey
from lib.src.types.py.block import Block
from lib.src.types.py.network import (
    send_message,
    receive_message,
    FetchTemplate,
    Template,
    ValidateTemplate,
    TemplateValidity,
    SubmitTemplate,
)

MINING_STEPS=10_000


class Miner:
    def __init__(self, reader: StreamReader, writer: StreamWriter, public_key: PublicKey,):
        self.public_key = public_key
        self.node_reader  = reader
        self.node_writer  = writer
        self._template_lock = Lock()  # shared state between async loop and mining thread
        self.mining = Event()
        self.current_template: Optional[Block] = None
        self.mined_blocks: queue.Queue[Block] = queue.Queue()
        self._start_mining_thread()

    @classmethod
    async def connect(cls, address: str, port: int, public_key: PublicKey) -> "Miner":
        reader, writer = await asyncio.open_connection(address, port)
        return cls(reader, writer, public_key)

    async def run(self) -> None:
        """ Main loop. """
        await asyncio.gather(self._template_loop(), self._submit_loop())

    async def _template_loop(self) -> None:
        while True:
            if not self.mining.is_set():
                # If not mining, fetch a new template block from the node
                await self._fetch_template()
            else:
                # If mining, check if the template block is still valid
                # - another miner already mined a block
                # - transactions change (double spend, replaced fee, ...)
                # - consensus may invalidate blocks after rules updates
                await self._validate_template()
            await asyncio.sleep(5)

    async def _submit_loop(self) -> None:
        loop = asyncio.get_running_loop()
        while True:
            # Wait for a mined block, then submit it
            block = await loop.run_in_executor(None, self.mined_blocks.get, True, None)
            await self._submit_block(block)

    def _start_mining_thread(self) -> None:
        """ Mining thread. """
        def worker():
            while True:
                self.mining.wait()  # Wait until receiving a request to start mining
                with self._template_lock:
                    template = self.current_template
                if template is None:
                    continue
                block = copy.deepcopy(template)
                print(f"Mining block with target: {block.header.target}")
                if block.header.mine(steps=MINING_STEPS):
                    print(f"Block mined: {block.hash()}")
                    self.mined_blocks.put(block)
                    self.mining.clear()
        Thread(target=worker, daemon=True).start()

    async def _fetch_template(self) -> None:
        print("Fetching new template")
        await send_message(self.node_writer, FetchTemplate(public_key=self.public_key))
        msg = await receive_message(self.node_reader)
        if not isinstance(msg, Template):
            raise RuntimeError("Unexpected message while fetching template")
        print(f"Received new template with target: {msg.block.header.target}")
        with self._template_lock:
            self.current_template = msg.block
        self.mining.set()  # Start!

    async def _validate_template(self) -> None:
        with self._template_lock:
            template = self.current_template
        if template is None:
            return
        await send_message(self.node_writer, ValidateTemplate(block=template), )
        msg = await receive_message(self.node_reader)
        if not isinstance(msg, TemplateValidity):
            raise RuntimeError("Unexpected message while validating template")
        if not msg.valid:
            print("Template invalidated")
            with self._template_lock:
                self.current_template = None
            self.mining.clear()  # Done!
        else:
            print("Template still valid")

    async def _submit_block(self, block: Block) -> None:
        print("Submitting new block!")
        await send_message(self.node_writer, SubmitTemplate(block=block))
        self.mining.clear()

File: miner/src/miner.rs
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::thread;
use std::time::Duration;
use anyhow::anyhow;
use flume::{Receiver, Sender};
use tokio::net::TcpStream;
use tokio::sync::Mutex;
use tokio::time::interval;
use btclib::crypto::PublicKey;
use btclib::network::Message;
use btclib::types::Block;
use log::{info, warn};

pub struct Miner {
    public_key: PublicKey,
    conn: Mutex<TcpStream>,
    current_template: Arc<std::sync::Mutex<Option<Block>>>,
    mining: Arc<AtomicBool>,
    mined_block_receiver: Receiver<Block>,
    mined_block_sender: Sender<Block>
}
impl Miner {
    pub(crate) async fn new(address: String, public_key: PublicKey) -> anyhow::Result<Self> {
        let stream = TcpStream::connect(&address).await?;
        let (mined_block_sender, mined_block_receiver) = flume::unbounded();
        info!("🔗 Connected to node: [{}]", address);
        Ok(Self {
            public_key,
            conn: Mutex::new(stream),
            current_template: Arc::new(std::sync::Mutex::new(None)),
            mining: Arc::new(AtomicBool::new(false)),
            mined_block_sender,
            mined_block_receiver
        })
    }

    pub(crate) async fn run(&self) -> anyhow::Result<()> {
        self.spawn_mining_thread();
        let mut template_interval = interval(Duration::from_secs(5));
        loop {
            let receiver_clone = self.mined_block_receiver.clone();
            tokio::select! {
                _ = template_interval.tick() => {
                    self.fetch_and_validate_template().await?;
                }
                Ok(mined_block) = receiver_clone.recv_async() => {
                    self.submit_block(mined_block).await?;
                }
            }
        }
    }

    fn spawn_mining_thread(&self) -> thread::JoinHandle<()> {
        let template = self.current_template.clone();
        let mining = self.mining.clone();
        let sender = self.mined_block_sender.clone();
        thread::spawn(move || loop {
            if mining.load(Ordering::Relaxed) {
                if let Some(mut block) = template.lock().unwrap().clone() {
                    if block.header.mine(2_000_000) {
                        info!("📦️Block mined: {}{}", " ".repeat(22), block.hash());
                        sender.send(block).expect("Failed to send mined block", );
                        mining.store(false, Ordering::Relaxed);
                    }
                }
            }
            thread::yield_now();
        })
    }

    async fn fetch_and_validate_template(&self) -> anyhow::Result<()> {
        if !self.mining.load(Ordering::Relaxed) {
            self.fetch_template().await?;
        } else {
            self.validate_template().await?;
        }
        Ok(())
    }

    async fn fetch_template(&self) -> anyhow::Result<()> {
        info!("Fetching new template");
        let message = Message::FetchTemplate(self.public_key.clone());
        let mut conn_lock = self.conn.lock().await;
        message.send_async(&mut *conn_lock).await?;
        drop(conn_lock);
        let mut conn_lock = self.conn.lock().await;
        match Message::receive_async(&mut *conn_lock).await? {
            Message::Template(template) => {
                drop(conn_lock);
                info!("↪️ Received new template with target: {}", template.header.target);
                *self.current_template.lock().unwrap() = Some(template);
                self.mining.store(true, Ordering::Relaxed);
                Ok(())
            }
            _ => Err(anyhow!("Unexpected message received when fetching template")),
        }
    }

    async fn validate_template(&self) -> anyhow::Result<()> {
        if let Some(template) = self.current_template.lock().unwrap().clone() {
            let message = Message::ValidateTemplate(template);
            let mut conn_lock = self.conn.lock().await;
            message.send_async(&mut *conn_lock).await?;
            drop(conn_lock);
            let mut conn_lock = self.conn.lock().await;
            match Message::receive_async(&mut *conn_lock).await? {
                Message::TemplateValidity(valid) => {
                    drop(conn_lock);
                    if !valid {
                        warn!("Current template is no longer valid");
                        self.mining.store(false, Ordering::Relaxed);
                    } else {
                        info!("Current template is still valid");
                    }
                    Ok(())
                }
                _ => Err(anyhow!("Unexpected message received when validating template")),
            }
        } else {
            Ok(())
        }
    }

    async fn submit_block(&self, block: Block) -> anyhow::Result<()> {
        info!("🚚 Submitting mined block");
        let message = Message::SubmitTemplate(block, self.public_key.clone());
        let mut conn_lock = self.conn.lock().await;
        message.send_async(&mut *conn_lock).await?;
        self.mining.store(false, Ordering::Relaxed);
        Ok(())
    }
}